sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+	ret = sof_ipc_tx_message_unlocked(sdev->ipc, rq.hdr.cmd,
+					  &rq, rq.hdr.size, partdata,
+					  sizeof(partdata->reply) + sizeof(u32));
+	if (ret < 0)
+		goto free;
+
+	sdev->next_comp_id = *(u32 *)partdata->data;
+
+	fw->data = vfe->tplg.data;
+	fw->pages = NULL;
+
+free:
+	mutex_unlock(&sdev->ipc->tx_mutex);
+
+	kfree(partdata);
+	return 0;
+}
+
+static int sof_vfe_sof_runtime_suspend(struct snd_sof_dev *sof_dev)
+{
+	return 0;
+}
+
+static int sof_vfe_sof_runtime_resume(struct snd_sof_dev *sof_dev)
+{
+	return 0;
+}
+
+static unsigned long get_dma_offset(struct snd_pcm_runtime *runtime,
+				    int channel, unsigned long hwoff)
+{
+	return hwoff + channel * (runtime->dma_bytes / runtime->channels);
+}
+
+static int sof_vfe_pcm_read_part(struct snd_sof_dev *sdev,
+				 struct snd_sof_pcm *spcm,
+				 struct snd_pcm_substream *substream,
+				 int channel, unsigned long pos,
+				 void __user *buf, unsigned long chunk_size)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	struct dsp_sof_data_resp *data = &vfe->data_resp;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
+	struct dsp_sof_data_req *req = (struct dsp_sof_data_req *)vfe->hdr_req;
+	DEFINE_SPINLOCK(lock);
+	unsigned int len;
+	int ret;
+
+	/* put response size in request */
+	req->size = chunk_size;
+	req->comp_id = spcm->stream[substream->stream].comp_id;
+	req->offset = get_dma_offset(substream->runtime, channel, pos);
+
+	sg_init_one(&sg_out, vfe->hdr_req, sizeof(vfe->hdr_req));
+	sg_init_one(&sg_in, data, chunk_size + HDR_SIZE_RESP);
+
+	ret = virtqueue_add_sgs(vfe->data_vq, sgs, 1, 1, vfe->hdr_req,
+				GFP_ATOMIC);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: could not send data %d\n", ret);
+
+	virtqueue_kick(vfe->data_vq);
+
+	ret = wait_for_completion_timeout(&vfe->completion,
+					  msecs_to_jiffies(DATA_TIMEOUT_MS));
+	if (!ret)
+		return -ETIMEDOUT;
+
+	virtqueue_disable_cb(vfe->data_vq);
+	spin_lock(&lock);
+
+	virtqueue_get_buf(vfe->data_vq, &len);
+
+	spin_unlock(&lock);
+	virtqueue_enable_cb(vfe->data_vq);
+
+	if (ret < 0)
+		return ret;
+	if (data->error < 0)
+		return data->error;
+
+	if (copy_to_user((void __user *)buf, data->data, chunk_size))
+		return -EFAULT;
+
+	return 0;
+}
+
+static int sof_vfe_pcm_write_part(struct snd_sof_dev *sdev,
+				  struct snd_sof_pcm *spcm,
+				  struct snd_pcm_substream *substream,
+				  int channel, unsigned long pos,
+				  void __user *buf, unsigned long chunk_size)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	struct dsp_sof_data_req *data = &vfe->data_req;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
+	struct dsp_sof_data_resp *resp = (struct dsp_sof_data_resp *)vfe->hdr_resp;
+	DEFINE_SPINLOCK(lock);
+	unsigned int len;
+	int ret;
+
+	data->size = chunk_size;
+	data->comp_id = spcm->stream[substream->stream].comp_id;
+	data->offset = get_dma_offset(substream->runtime, channel, pos);
+
+	if (copy_from_user(data->data, (void __user *)buf, chunk_size))
+		return -EFAULT;
```

Comment:
that doesn't look good to me.
You should have a ring buffer for the guest, then when you update the position you use the .ack to copy the data over to the host.
And in a second step you make all this transparent and the same buffer is shared between host and guest.

Reply:
@plbossart sorry, not sure what you mean. This is the playback path. Here the audio data is copied from a user-space process to the VirtQueue to be sent to the host. Maybe the comment is now displayed with a wrong context, so I'm not sure what you're referring to.

Reply:
@lyakh ALSA does the copy from user space for you. You should only need to copy into the virtio ringbuffer from the ALSA buffer (when not using DMA copy mode).

Reply:
@lgirdwood you mean the default `default_write_copy()` / `default_read_copy()` implementation, right? But if I keep that, how do I get signalled when that copy is called, i.e. when the user either sends data for playback or supplies a buffer for capture? But even if I get that, it would mean, that on the one hand `default_{read,write}_copy()` copy data to / from an ALSA buffer, additionally I also have to copy data between that buffer and VirtQ? So, I think I would have an additional memcpy(), even if it's "just" an in-kernel copy, no user-space, no IO, etc.

Reply:
@lyakh The data flow should be
```
G = Guest.
H = Host (access to real HW).
G aplay -> G ALSA buffer -> G drv copies to virtio buffer -> H virtio buffer -> Host buffer -> DSP
```
Can you confirm.

Reply:
@lgirdwood roughly - yes, subject to the exact meaning of "Host buffer" - is that supposed to be an ASLA buffer or just some buffer on the host?

Reply:
It should be a DMAable buffer on the host. NOT an ALSA buffer as there is no host userspace IO to this buffer.

Reply:
@lgirdwood right, I think that's exactly what is happening in `dsp_sof_stream_playback()`:
```
	memcpy(runtime->dma_area + req->offset, req->data, data_size);
```
and in `dsp_sof_stream_capture()`
```
	memcpy(reply->data, runtime->dma_area + req->offset, data_size);
```